Geometry/HGCalCommonData/plugins/dd4hep/DDHGCalEEFileAlgo.cc

```diff
@@ -0,0 +1,402 @@
+        }
+      }
+    }  // End of loop over blocks
+
+#ifdef EDM_ML_DEBUG
+    edm::LogVerbatim("HGCalGeom") << "DDHGCalEEFileAlgo: " << copies.size() << " different wafer copy numbers";
+    int k(0);
+    for (std::unordered_set<int>::const_iterator itr = copies.begin(); itr != copies.end(); ++itr, ++k) {
+      edm::LogVerbatim("HGCalGeom") << "Copy [" << k << "] : " << (*itr);
+    }
+    copies.clear();
+    edm::LogVerbatim("HGCalGeom") << "<<== End of DDHGCalEEFileAlgo construction...";
+#endif
+  }
+
+  void positionSensitive(cms::DDParsingContext& ctxt,
+                         xml_h e,
+                         const dd4hep::Volume& glog,
+                         double rin,
+                         double rout,
+                         double zpos,
+                         int layertype,
+                         int layer) {
+    cms::DDNamespace ns(ctxt, e, true);
+    static const double sqrt3 = std::sqrt(3.0);
+    int layercenter = layerCenter[layer];
+    double r = 0.5 * (waferSize + waferSepar);
+    double R = 2.0 * r / sqrt3;
+    double dy = 0.75 * R;
+    int N = (int)(0.5 * rout / r) + 2;
+    const auto& xyoff = geomTools.shiftXY(layercenter, (waferSize + waferSepar));
+#ifdef EDM_ML_DEBUG
+    int ium(0), ivm(0), iumAll(0), ivmAll(0), kount(0), ntot(0), nin(0);
+    std::vector<int> ntype(6, 0);
+    edm::LogVerbatim("HGCalGeom") << "DDHGCalEEFileAlgo: " << glog.name() << " rin:rout " << rin << ":" << rout
+                                  << " zpos " << zpos << " N " << N << " for maximum u, v;  r " << r << " R " << R
+                                  << " dy " << dy << " Shift " << xyoff.first << ":" << xyoff.second << " WaferSize "
+                                  << (waferSize + waferSepar);
+#endif
+    for (int u = -N; u <= N; ++u) {
+      int iu = std::abs(u);
+      for (int v = -N; v <= N; ++v) {
+        int iv = std::abs(v);
+        int nr = 2 * v;
+        int nc = -2 * u + v;
+        double xpos = xyoff.first + nc * r;
+        double ypos = xyoff.second + nr * dy;
+        const auto& corner = HGCalGeomTools::waferCorner(xpos, ypos, r, R, rin, rout, false);
+#ifdef EDM_ML_DEBUG
+        ++ntot;
+        if (((corner.first <= 0) && std::abs(u) < 5 && std::abs(v) < 5) || (std::abs(u) < 2 && std::abs(v) < 2)) {
+          edm::LogVerbatim("HGCalGeom") << "DDHGCalEEFileAlgo: " << glog.name() << " R " << rin << ":" << rout
+                                        << "\n Z " << zpos << " LayerType " << layertype << " u " << u << " v " << v
+                                        << " with " << corner.first << " corners";
+        }
+#endif
+        int type = HGCalWaferType::getType(HGCalWaferIndex::waferIndex(layer, u, v, false), waferIndex, waferTypes);
+        if (corner.first > 0 && type >= 0) {
+          int copy = type * 1000000 + iv * 100 + iu;
+          if (u < 0)
+            copy += 10000;
+          if (v < 0)
+            copy += 100000;
```

Comment:
As mentioned in the previous PR, please make these magic numbers (1e4, 1e5, 1e6) into constants.